build.sc

```diff
@@ -180,9 +180,17 @@ trait AmmInternalModule extends CrossSbtModule with Bloop.Module{
     acyclicOptions ++ tastyReaderOptions
   }
   def compileIvyDeps = T {
-    if (isScala2()) Agg(Deps.acyclic)
-    else Agg[Dep]()
+    Agg(Deps.mainargs) ++ (
+      if (isScala2()) Agg(Deps.acyclic)
+      else Agg[Dep]()
+    )
   }
+  def runIvyDeps =
+    if (scala3Versions.contains(crossScalaVersion))
+      Agg(ivy"com.lihaoyi:mainargs_3:${Deps.mainargs.dep.version}")
+    else
+      Agg(Deps.mainargs)
```

Comment:
Interesting approach!
So you compile with the `_2.13` artifact and then at runtime you change it with the `_3` so Scala 3 macros are available for users to compile their code.
If this works here, we could use the same approach to fix upickle which currently works in Scala 3 only in `--thin` mode + `$ivy` import but not normally since we have the `_2.13` artifact at runtime.

Reply:
I think you might wat to add a comment before `runIvyDeps` to explain what you are doing. It is not trivial to understand and people might get confused in the future.

Reply:
> So you compile with the _2.13 artifact and then at runtime you change it with the _3 so Scala 3 macros are available for users to compile their code.

Yes exactly - and yes you're right that totally deserves a comment - added just now. 
As far as I'm concerned this approach works, and it may also be the way to fix the messed-up dependency tree - see e.g. https://github.com/com-lihaoyi/Ammonite/issues/1241.